components/brave_rewards/renderer/creator_detection_agent.cc

```diff
@@ -0,0 +1,265 @@
+  DCHECK(web_frame);
+  if (web_frame->IsProvisional()) {
+    return std::nullopt;
+  }
+
+  // Only run scripts for known "media platform" sites, and only over HTTPS.
+  GURL origin = url::Origin(web_frame->GetSecurityOrigin()).GetURL();
+  if (!origin.is_valid() || !origin.SchemeIs(url::kHttpsScheme) ||
+      !IsMediaPlatformURL(origin)) {
+    return std::nullopt;
+  }
+
+  auto iter = kScriptMap.find(origin.host_piece());
+  if (iter == kScriptMap.end()) {
+    return std::nullopt;
+  }
+
+  return DetectionScript{
+      .source = blink::Platform::Current()->GetDataResourceString(iter->second),
+      .should_run_in_isolated_world = ShouldRunInIsolatedWorld(iter->second)};
+}
+
+v8::MaybeLocal<v8::Function> CompileDetectionInitializer(
+    v8::Local<v8::Context> context,
+    std::string_view function_body) {
+  v8::Context::Scope context_scope(context);
+  v8::MicrotasksScope microtasks_scope(
+      context, v8::MicrotasksScope::kDoNotRunMicrotasks);
+
+  auto* isolate = context->GetIsolate();
+  v8::ScriptCompiler::Source source(gin::StringToV8(isolate, function_body));
+
+  v8::Local<v8::String> parameters[] = {
+      gin::StringToV8(isolate, "setPageChangedCallback"),
+      gin::StringToV8(isolate, "onCreatorDetected"),
+      gin::StringToV8(isolate, "verboseLogging")};
+
+  // The creator detection script is compiled as a function body. Browser
+  // capabilities are passed in to the compiled function as arguments whose
+  // parameter names are defined above. Browser capabilities are only exposed
+  // to the script; they are not visible outside of the compiled function.
+  return v8::ScriptCompiler::CompileFunction(context, &source, 3, parameters);
+}
+
+template <typename... Args>
+void CallFunctionWithArgs(v8::Isolate* isolate,
+                          v8::Local<v8::Function> function,
+                          Args... args) {
+  v8::Local<v8::Context> context = function->GetCreationContextChecked(isolate);
+  v8::Context::Scope context_scope(context);
+  v8::MicrotasksScope microtasks_scope(
+      context, v8::MicrotasksScope::kDoNotRunMicrotasks);
+  std::array<v8::Local<v8::Value>, sizeof...(Args)> arg_list = {
+      args.template As<v8::Value>()...};
+  std::ignore = function->Call(context, context->Global(), arg_list.size(),
+                               arg_list.empty() ? nullptr : arg_list.data());
+}
+
+}  // namespace
+
+CreatorDetectionAgent::CreatorDetectionAgent(content::RenderFrame* render_frame,
+                                             int32_t isolated_world_id)
+    : RenderFrameObserver(render_frame),
+      isolated_world_id_(isolated_world_id) {}
+
+CreatorDetectionAgent::~CreatorDetectionAgent() = default;
+
+void CreatorDetectionAgent::DidCommitProvisionalLoad(
+    ui::PageTransition transition) {
+  if (render_frame()->IsMainFrame()) {
+    MaybeInjectDetectionScript();
+  }
+}
+
+void CreatorDetectionAgent::DidFinishSameDocumentNavigation() {
+  if (render_frame()->IsMainFrame()) {
```

Comment:
I've remove the check here, since there's a `page_changed_callback_.IsEmpty()` check in `MaybeNotifyPageChanged`, and that callback only gets set for the main frame. Am I understanding that correctly?